src/snowflake/snowpark/mock/_functions.py

```diff
@@ -304,34 +306,66 @@ def mock_to_date(
 
         [x] If the variant contains a string, a string conversion is performed.
 
-        [ ] If the variant contains a date, the date value is preserved as is.
+        [x] If the variant contains a date, the date value is preserved as is.
 
-        [ ] If the variant contains a JSON null value, the output is NULL.
+        [x] If the variant contains a JSON null value, the output is NULL.
 
         [x] For NULL input, the output is NULL.
 
-        [ ] For all other values, a conversion error is generated.
+        [x] For all other values, a conversion error is generated.
     """
     res = []
-    auto_detect = bool(not fmt)
 
-    date_format, _, _ = convert_snowflake_datetime_format(
-        fmt, default_format="%Y-%m-%d"
-    )
+    if not isinstance(fmt, ColumnEmulator):
+        fmt = [fmt] * len(column)
+
+    for data, _fmt in zip(column, fmt):
+        auto_detect = _fmt is None or _fmt.lower() == "auto"
+
+        date_format, _, _ = convert_snowflake_datetime_format(
+            _fmt, default_format="%Y-%m-%d"
+        )
+        import dateutil.parser
 
-    for data in column:
```

Comment:
NIT: we should be able to easily rewrite the for loop to use `Series.apply` and `try_convert` to be consistent with the rest of the casting functions.

Reply:
refactored